### PR TITLE
kubelet: Refactor tryUpdateNodeStatus() into smaller functions

### PR DESCRIPTION
**Pull request by jingyuanliang**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Support #112618. This replaces #113188.

#### Which issue(s) this PR fixes:

This breaks down tryUpdateNodeStatus into multiple functions.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -478,21 +478,39 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
 	}
-
-	originalNode := node.DeepCopy()
 	if originalNode == nil {
 		return fmt.Errorf("nil %q node object", kl.nodeName)
 	}
 
+	node, changed := kl.updateNode(originalNode)
+
+	if kl.clock.Since(kl.lastStatusReportTime) < kl.nodeStatusReportFrequency && !changed {
```
[the remark this answers is not in the document]
Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:491]: ah, no, it is ` !podCIDRChanged && !nodeStatusHasChanged(&originalNode.Status, &node.Status) && !areRequiredLabelsNotPresent`

File: pkg/kubelet/kubelet_node_status.go [hunk 2]
```diff
@@ -478,21 +478,39 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
+		kl.markVolumesFromNode(node)
+		return nil
+	}
+
+	updatedNode, err := kl.patchNode(originalNode, node)
+	if err == nil {
+		kl.markVolumesFromNode(updatedNode)
```
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:498]:
```suggestion
	// If update finishes successfully, mark the volumeInUse as reportedInUse to indicate
 	// those volumes are already updated in the node's status
		kl.markVolumesFromNode(updatedNode)
```

I suggest to keep the comments

File: pkg/kubelet/kubelet_node_status.go [hunk 3]
```diff
@@ -520,41 +538,47 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
...
-			kl.volumeManager.MarkVolumesAsReportedInUse(node.Status.VolumesInUse)
-			return nil
-		}
-	}
+	return node, podCIDRChanged || nodeStatusHasChanged(&originalNode.Status, &node.Status) || areRequiredLabelsNotPresent
+}
 
+// patchNode patches node on the API server based on originalNode.
+// It returns any potential error, or an updatedNode and refreshes the state of kubelet when successful.
+func (kl *Kubelet) patchNode(originalNode, node *v1.Node) (*v1.Node, error) {
```
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:546]: it should be `patchNodeStatus`

**Comment by aojea** [comment]
so, right now it seems that the update node process is undocumented, if we break it down and we start to use independent fragments I'm not sure we'll respect the kubelet update node process now or in the future unless we create a test that defines this process

File: pkg/kubelet/kubelet_node_status.go (hunk 2 above)
Comment (jingyuanliang) [reply pkg/kubelet/kubelet_node_status.go:498]: This comment was moved to the top of markVolumesFromNode.

**Comment by jingyuanliang** [comment]
Is the current tryUpdateNodeStatus() already covered by tests?

File: pkg/kubelet/kubelet_node_status.go (hunk 2 above)
Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:498]: oh, thanks, sorry

**Comment by aojea** [comment]
> Is the current tryUpdateNodeStatus() already covered by sufficient amount of tests?

I don't think we have good tests in this area, I remember one regression that increased the time to start kubelet https://github.com/kubernetes/kubernetes/pull/99336 and went undetected

I really think we should do this, but I'm scared of regressions ...

What is your intent with this patch, I assume that this is part of a larger change, can you add subsequent commits so we can review the whole solution proposed?

**Comment by jingyuanliang** [comment]
https://github.com/kubernetes/kubernetes/pull/112618 was updated to show the use with this refactor. Let's keep this PR open so we can run tests on the refactor alone.

**Comment by aojea** [comment]
/close

superseded by https://github.com/kubernetes/kubernetes/pull/112618

**Pull request closed without merging**